### PR TITLE
Transit Spice for Switzerland

### PR DESCRIPTION
**Pull request by tagawa**
**What does your Instant Answer do?**
Shows the next train times for trains in Switzerland (and some major European cities).

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It highlights key information - departure time, number of transfers, etc. - for multiple trains at a time, in a single page.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://transport.opendata.ch

**Why did you choose this data source?**
It's free and easy to work with.

**Are there any other alternative (better) data sources?**
Not that I'm aware of.

**What are some example queries that trigger this Instant Answer?**
"next train to geneva from paris"
"train times zurich airport to basel"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
People living in or visiting Switzerland

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
https://duck.co/ideas/idea/738/train-departures-show-train-info

**Which existing Instant Answers will this one supersede/overlap with?**
Will work side-by-side other transit IAs. 

**Are you having any problems? Do you need our help with anything?**
I'd like to add the ability for search terms to have accented characters.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
The API is free to use with no API key required.

**Where did you hear about DuckDuckHack? (For first time contributors)**
N/A

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
ddg_spice_transit_switzerland

## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Ubuntu 14
        [X] Chromium
        [X] Firefox
        [X] Opera
        [X] Midori (WebKit)

    - Android
        [X] Firefox
        [X] Native Browser
        [] Google Chrome
        [X] Opera
```

### CODE REVIEW

**Comment by moollaza** [comment]
@tagawa this is awesome! We'll be sure to review this shortly and give you some feedback!

**Comment by tagawa** [comment]
Great, thank you. It should work as it is but for a future update I'd like to learn how to handle accented characters such as in Zürich, Genève, etc.

**Comment by mattr555** [comment]
\o/ Awesome! It's pretty cool that Switzerland has an open API and we don't have to resort to building our own...

+ Would it be advantageous to display the train number [like we do for NJT?](https://duckduckgo.com/?q=next+train+from+metropark+to+new+york+penn&ia=njtransit) I'm obviously not familiar with the Swiss system but it could be useful to corroborate the results with station signage. If we decide to do that, we could move the platform # below the status indicator.
+ Probably not as important for an initial release, but since we have transfer info, can we show it using a item_detail expander (i.e. one like the [recipe IA](https://duckduckgo.com/?q=chocolate+cookie+recipe&ia=recipes&iai=0))? Perhaps @chrismorast has an idea for its design.
+ I noticed that it doesn't look great on the dark mode, but neither do the other transit IAs. We can use a (currently undocumented) class that automatically updates the color based on the theme.
screenshot - 03052015 - 01 16 20 pm

Thanks!

File: share/spice/transit/switzerland/transit_switzerland.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                    moreAt: true
+                }
+            },
+            normalize: function(item){
+                return {
+                    arrival_time: format_time(item.to.arrival),
+                    departure_time: format_time(item.from.departure),
+                    platform: item.from.platform,
+                    status: (item.from.delay) ? 'Delayed' : 'On time',
+                    status_class: (item.from.delay) ? 'septa__delayed' : '',
```
Comment (mattr555) [inline share/spice/transit/switzerland/transit_switzerland.js]: s/septa/switzerland

File: share/spice/transit/switzerland/transit_switzerland.css [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+.tile--switzerland .switzerland__time {
+    font-size: 2.3em;
+    color: #333;
+    font-weight: 300;
+}
+
+.tile--switzerland .switzerland__details {
+    color: #999;
+    margin-top: 5px;
+    margin-bottom: 5px;
```
Comment (mattr555) [inline share/spice/transit/switzerland/transit_switzerland.css]: Nitpicking, but could we make this 10px to correspond with the other answers?

**Comment by tagawa** [comment]
Thanks for the fast review @mattr555. Please do nitpick!

* I've found where train numbers are in the API so I'll add that.
* Good idea about the transfer details in ```item_detail```. I'll put that as a todo for an update.
* Dark mode - I forgot about that! I'll fix that and the remaining comments.

**Comment by tagawa** [comment]
ddg_spice_transit_switzerland_dark

Looks better now. If these changes are OK, I can update the other transit IAs for consistency.

**Comment by mattr555** [comment]
Hi @tagawa, forgive me for not being more specific, but I think the size of the times were ok. We can keep `.text--primary` but can probably lose `.t-xxxxl` and revert back to the sizes we had earlier. Likewise, the color of the status indicator may not need to be changed. (we spent a while going back and forth on those :smile:)

Looks good on mobile :+1: 

mobile

Also see #1600 :)

**Comment by tagawa** [comment]
Hi @mattr555. Thanks for the explanation - just made those two reversions. I was playing a bit too safe using only the colours and sizes in the style guide.

I'll have a go at #1600 now.

**Comment by moollaza** [comment]
Ping! Hey guys, what's the status here? Ready for a code/design review?

**Comment by tagawa** [comment]
If @mattr555 agrees I think it's ready for code/design review.
#1600 is a separate issue and is closed.

File: lib/DDG/Spice/Transit/Switzerland.pm [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+spice wrap_jsonp_callback => 1;
+spice proxy_cache_valid => "418 1d";
+
+#load a list of stops so we don't trigger this if we don't get train stops
+#(the triggers are similar to SEPTA's)
+my @stops = share('stops.txt')->slurp;
+
+#check if the stop name is in the list of stops
+#(using the same matching algorithm as the backend)
+sub is_stop {
```
Comment (moollaza) [inline lib/DDG/Spice/Transit/Switzerland.pm]: This is pretty inefficient. We've been using YAML files in lots of Goodies with `use YAML::XS qw(load)`

If we convert stops.text to valid YAML it means we'll be able to slurp the triggers into a hash. That way we can use a hash lookup to check if the candidate location is valid. Much faster than looping over 2000+ stops!

Comment (moollaza) [inline lib/DDG/Spice/Transit/Switzerland.pm]: Lots of examples: https://github.com/duckduckgo/zeroclickinfo-goodies/search?utf8=%E2%9C%93&q=YAML

File: lib/DDG/Spice/Transit/Switzerland.pm [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    foreach my $stop (@stops){
+        return 1 if index(lc $stop, lc $_[0]) > -1;
+    }
+    return;
+};
+
+triggers any => "next train", "train times", "train schedule";
```
Comment (moollaza) [inline lib/DDG/Spice/Transit/Switzerland.pm]: Should probably be `triggers startend` -- all our expected queries don't have these triggers in the middle of the query string.

If we wanted to support queries like `when is the next train from .. to ...` then we should keep this and add a regex to strip filler words in the `handle`

File: share/spice/transit/switzerland/transit_switzerland.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        
+        var to = api_result.to.name || '',
+            from = api_result.from.name || '';
+        
+        Spice.add({
+            id: "switzerland",
+            name: "Swiss Trains",
+            data: api_result.connections,
+            meta: {
+                heading: from + " to " + to,
```
Comment (moollaza) [inline share/spice/transit/switzerland/transit_switzerland.js]: this can be removed -- we don't have a `heading` property for `meta`

**Comment by chrismorast** [comment]
@mattr555 in regards to https://github.com/duckduckgo/zeroclickinfo-spice/pull/1593#issuecomment-77419802  (sorry, just saw that); the ideal scenario would be to open up a map with highlighted route - which is the functionality we're working on getting implemented and hope to add to other transit IAs you did as well.

**Comment by tagawa** [comment]
Thanks for the review @moollaza and the optimization advice. I've changed it to do a hash lookup, although because the YAML content is a 1-dimensional array, I wonder if ```if (grep {$_ eq $_[0]} @stops)``` would be more performant. What do you recommend?

**Comment by moollaza** [comment]
@tagawa I think it would be simplest to change the YAML from an array to a hash like [stops.yml](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/transit/septa/stops.yml)

Map the locations to a `~` or `1` and then you magically have a hash you can use for a simplified lookup :)

**Comment by tagawa** [comment]
Good idea @moollaza. Now done and it seems to work well.

**Comment by tagawa** [comment]
Thanks @moollaza - done.

File: share/spice/transit/switzerland/transit_switzerland.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+(function (env) {
+    "use strict";
+    env.ddg_spice_transit_switzerland = function(api_result){
+
+        if (api_result.error || !api_result.to || !api_result.from || !api_result.connections) {
+            return Spice.failed('transit_switzerland');
```
Comment (mattr555) [inline share/spice/transit/switzerland/transit_switzerland.js]: The argument here should match the `id` field of your `Spice.add()` (so change it to just `'switzerland'`)

File: share/spice/transit/switzerland/transit_switzerland.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
+(function (env) {
+    "use strict";
+    env.ddg_spice_transit_switzerland = function(api_result){
+
+        if (api_result.error || !api_result.to || !api_result.from || !api_result.connections) {
```
Comment (mattr555) [inline share/spice/transit/switzerland/transit_switzerland.js]: We like to check `api_result` exists as well just to be safe.

File: share/spice/transit/switzerland/transit_switzerland.js (hunk 6 above)
Comment (tagawa) [inline share/spice/transit/switzerland/transit_switzerland.js]: Thanks. Just changed this and added the check for ```api_result```

**Comment by mattr555** [comment]
@tagawa Wow, you're fast :smiley:

Everything looks good to me :star:
Works on IE8 (except for `border-radius` -- no big deal):
screenshot - 04032015 - 10 01 20 pm

FWIW, I created #1728 to speed up our stop searches on the other transit answers.

**Comment by tagawa** [comment]
Thanks for the review and earlier feedback. Good idea about #1728 &mdash; I'll look into that separately.

**Comment by jagtalon** [comment]
Looks like all is good here. @abeyang no need for design here since it looks exactly like our other IAs such as SEPTA: https://duckduckgo.com/?q=next+train+from+strafford+to+paoli&t=ffab&ia=septa.

@mattr555 if you're comfortable, please merge. :)

**Comment by mattr555** [comment]
Ooh, my first merge of a brand new IA :)

Thanks a lot @tagawa! :cake:

**Comment by tagawa** [comment]
Thanks and congratulations! :tada:

**Comment by moollaza** [comment]
@tagawa congrats! It's live: https://duckduckgo.com/?q=next+train+to+geneva+from+paris&ia=swisstrains

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

**Comment by tagawa** [comment]
Thank you! Great to see this in the wild and available to users.

I'd like to find more transit APIs to make IAs from, perhaps having a shared directory for common transit files (e.g. CSS files).

**Pull request merged**